### DISCUSSION TITLE
Monaco - Pyright LSP: runtime modules based on user changes

### DISCUSSION THREAD

**Discussion by hugoribeiromsglife** [microsoft/pyright#7123] [Q&A]
Hey,
 
First and foremost, let me share with you that I am working with LSPs for the first time, therefore these past weeks were basically spent reading, trying to understand and learn something about LSPs and giving a bunch of things a try.
 
I already have my webapp using monaco as a language server client connecting to pyright as a server and everything seems to be working as expected (for pyright's python modules).
 
The challenge I am facing and currently googling, trying to find some possible solutions is the following (I do not even know if the scenario makes sense but is what I need to somehow implement):
 
1. we have one pyright server running.
2. the webapp will be used by a lot of users at the same time and they will be changing our app model dinamically/runtime and I would like to make that dynamic data known by the server so that I am able to be suggested/validated accordingly. e.g. user 1 is using the webapp and creates a new function 'abc' (outside the editor) when this user 1 opens the monaco editor can we make pyright know this function abc (can make it a python def if needed be)? if at the same time a user 2 is using the app and creates a function 'def' (outside the editor) this one would need to be known by the server when that user opens the editor on his browser. If there is a way to do it without creating files in the filesystem, that would be nice.
 
I must admit I am new to this and a bit lost.. That is why I ended up deciding to share this question with you.
 
Thanks in advance, Best regards,
Hugo.

**Comment by erictraut** [comment]
You may find it useful to look at the code for my "pyright playground" website. Here's the github site: https://github.com/erictraut/pyright-playground. Here's the running web app: https://pyright-play.net/. It sounds like you are trying to do something similar, so perhaps you can leverage the same techniques.

**Reply by hugoribeiromsglife** [reply]
Will do that, thanks again!

**Reply by hugoribeiromsglife** [reply]
I have been taking a look at pyright playground's code and if I am not mistaken there you can have multiple sessions/clients/models being those new modules/functions written inside the monaco editor, right? I want to add extra modules/functions at runtime but those are not written inside the editor e.g. before opening the editor I would need to make an http request to my BE, grab the response which can be a string containing my py module definitions and I would like to on the fly inject those modules on that user's pyright workspace so that the server can recognize/validate them.
 
Is this possible somehow?

before opening the editor I would get the py content from the BE, create an in memory file (not in the file system):
![image](https://github.com/microsoft/pyright/assets/88100226/5ae81bfc-1572-40bb-9160-81f1d26414c5)

the monaco editor is recognizing it:
![image](https://github.com/microsoft/pyright/assets/88100226/418d4165-56ef-4dde-9d25-27810eab5381)

but then pyright does not know about them:
![image](https://github.com/microsoft/pyright/assets/88100226/5fa36ac1-60bf-4934-a417-36dc2adc67a3)

**Reply by hugoribeiromsglife** [reply]
Hi again,

Ended up being able to implement what I needed by creating different workspaces per each user/client on server-side and letting each user/client know which workspace should he/she point to.

Thanks!
